src/PsrLogAdapter.php

```diff
@@ -0,0 +1,118 @@
+<?php
+/*
+ * Copyright 2023-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace MongoDB;
+
+use MongoDB\Driver\Logging\Logger as DriverLogger;
+use MongoDB\Exception\UnexpectedValueException;
+use Psr\Log\LoggerInterface as PsrLogger;
+use Psr\Log\LogLevel as PsrLogLevel;
+use SplObjectStorage;
+
+use function MongoDB\Driver\Logging\addLogger;
+use function MongoDB\Driver\Logging\removeLogger;
+use function sprintf;
+
+final class PsrLogAdapter implements DriverLogger
+{
+    private static ?self $instance = null;
+
+    /** @psalm-var SplObjectStorage<PsrLogger, null> */
+    private SplObjectStorage $psrLoggers;
+
+    private const PSR_LEVELS = [
+        DriverLogger::LEVEL_ERROR => PsrLogLevel::ERROR,
+        /* libmongoc considers "critical" less severe than "error" so map it to
+         * "error" in the PSR logger. */
+        DriverLogger::LEVEL_CRITICAL => PsrLogLevel::ERROR,
+        DriverLogger::LEVEL_WARNING => PsrLogLevel::WARNING,
+        DriverLogger::LEVEL_MESSAGE => PsrLogLevel::NOTICE,
+        DriverLogger::LEVEL_INFO => PsrLogLevel::INFO,
+        DriverLogger::LEVEL_DEBUG => PsrLogLevel::DEBUG,
+    ];
+
+    public static function addLogger(PsrLogger $psrLogger): void
+    {
+        $instance = self::getInstance();
+
+        $instance->psrLoggers->attach($psrLogger);
+
+        addLogger($instance);
+    }
+
+    /**
+     * Pipes a log message from PHPC to all registered PSR loggers.
+     *
+     * @internal
+     * @see DriverLogger::log()
+     */
+    public function log(int $level, string $domain, string $message): void
+    {
+        if (! isset(self::PSR_LEVELS[$level])) {
+            throw new UnexpectedValueException(sprintf(
+                'Expected level to be >= %d and <= %d, %d given for domain "%s" and message: %s',
+                DriverLogger::LEVEL_ERROR,
+                DriverLogger::LEVEL_DEBUG,
+                $level,
+                $domain,
+                $message,
+            ));
+        }
+
+        $instance = self::getInstance();
+        $psrLevel = self::PSR_LEVELS[$level];
+        $context = ['domain' => $domain];
+
+        foreach ($instance->psrLoggers as $psrLogger) {
+            $psrLogger->log($psrLevel, $message, $context);
+        }
+    }
+
+    public static function removeLogger(PsrLogger $psrLogger): void
```

Comment:
removeLogger should be just after addLogger because they are related. And internal methods should be last (as privates).

Reply:
Spoke to @alcaeus about this. I'll add a skip for PsrLogAdapter in PedantryTest for now and re-order these methods. We can revisit removing PedantryTest altogether in a subsequent PR.

Reply:
I found it was easier to just add `@internal` to the sort logic. Opened [PHPLIB-1238](https://jira.mongodb.org/browse/PHPLIB-1238) to track this (already implemented locally) and will address in a subsequent PR.

Reply:
#1162 is the PR for method sorting changes. If that gets merged first I'll update this PR to be consistent before merging.